toolbox/include/base/Parameters.h

```diff
@@ -0,0 +1,265 @@
+/*
+ * Copyright (C) 2018 Istituto Italiano di Tecnologia (IIT)
+ * All rights reserved.
+ *
+ * This software may be modified and distributed under the terms of the
+ * GNU Lesser General Public License v2.1 or any later version.
+ */
+
+#ifndef WBT_PARAMETERS_H
+#define WBT_PARAMETERS_H
+
+#include <memory>
+#include <string>
+#include <vector>
+
+namespace wbt {
+    template <typename T>
+    class Parameter;
+    class ParameterMetadata;
+    class Parameters;
+    const int PARAM_INVALID_INDEX = -1;
+    const std::string PARAM_INVALID_NAME = {};
+} // namespace wbt
+
+/**
+ * @brief Class for storing block's parameters
+ *
+ * This class can contain scalar and vector parameters of the supported types.
+ *
+ * @see wbt::Parameter, wbt::ParameterMetadata, wbt::ParameterType
+ */
+class wbt::Parameters
+{
+public:
+    using ParamIndex = int;
+    using ParamName = std::string;
+
+private:
+    class impl;
+    std::unique_ptr<impl, void (*)(impl*)> pImpl;
+
+public:
+    Parameters();
+    Parameters(const wbt::Parameters& other);
+    Parameters& operator=(const Parameters& other);
+
+    ~Parameters() = default;
+
+    /**
+     * @brief Get the name of a stored parameter from its index.
+     * @param index The index of the parameter.
+     * @return The name if the parameter exists, wbt::PARAM_INVALID_NAME otherwise.
+     */
+    ParamName getParamName(const ParamIndex& index) const;
+
+    /**
+     * @brief Get the index of a stored parameter from its name.
+     * @param name The name of the parameter.
+     * @return The index if the parameter exists, wbt::PARAM_INVALID_INDEX otherwise.
+     */
+    ParamIndex getParamIndex(const ParamName& name) const;
+
+    /**
+     * @brief Get the number of stored parameters
+     *
+     * @return The number of stored parameters.
+     */
+    unsigned getNumberOfParameters() const;
+
+    /**
+     * @brief Check if a parameter with a given name is stored
+     *
+     * @param name The name of the parameter.
+     * @return True if the parameters exists, false otherwise.
+     */
+    bool existName(const ParamName& name) const;
+
+    /**
+     * @brief Store a scalar parameter
+     *
+     * @tparam The type of the parameter to store. Despite this, the parameter get cast accordingly
+     *         to its metadata.
+     * @param param The value of the scalar parameter to store.
+     * @param paramMetadata The metadata associated to the parameter to store.
+     * @return True for success, false otherwise.
+     */
+    template <typename T>
+    bool storeParameter(const T& param, const wbt::ParameterMetadata& paramMetadata);
```

Comment:
Have you tried compiling and linking this on Windows? I think the `WINDOWS_EXPORT_ALL_SYMBOLS` option will be quite stressed by this edge case.

Reply:
Also giving it a shot on macOS may be worth, given all the intensive use of templates.
You can use my former iMac (the idea is using it for that). cc @DanielePucci

Reply:
> Have you tried compiling and linking this on Windows?

Of course no :grin: We actually claim multiplatform support but we never tested broadly on non-linux platforms. For this reason some while ago I decided to put a [big disclaimer](https://robotology.github.io/wb-toolbox/mkdocs/install/#install) on the website.

Reply:
No, apparently we cannot use the iMac because it does not have a Matlab/Simulink license.